src/main/java/rx/internal/operators/OnSubscribeFromAsync.java

```diff
@@ -0,0 +1,514 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package rx.internal.operators;
+
+import java.util.Queue;
+import java.util.concurrent.atomic.*;
+
+import rx.*;
+import rx.Observable.OnSubscribe;
+import rx.exceptions.*;
+import rx.functions.Action1;
+import rx.internal.util.RxRingBuffer;
+import rx.internal.util.atomic.SpscUnboundedAtomicArrayQueue;
+import rx.internal.util.unsafe.*;
+import rx.plugins.RxJavaHooks;
+import rx.subscriptions.SerialSubscription;
+
+public final class OnSubscribeFromAsync<T> implements OnSubscribe<T> {
+
+    final Action1<AsyncEmitter<T>> asyncEmitter;
+    
+    final AsyncEmitter.BackpressureMode backpressure;
+    
+    public OnSubscribeFromAsync(Action1<AsyncEmitter<T>> asyncEmitter, AsyncEmitter.BackpressureMode backpressure) {
+        this.asyncEmitter = asyncEmitter;
+        this.backpressure = backpressure;
+    }
+    
+    @Override
+    public void call(Subscriber<? super T> t) {
+        BaseAsyncEmitter<T> emitter;
+        
+        switch (backpressure) {
+        case NONE: {
+            emitter = new NoneAsyncEmitter<T>(t);
+            break;
+        }
+        case ERROR: {
+            emitter = new ErrorAsyncEmitter<T>(t);
+            break;
+        }
+        case DROP: {
+            emitter = new DropAsyncEmitter<T>(t);
+            break;
+        }
+        case LATEST: {
+            emitter = new LatestAsyncEmitter<T>(t);
+            break;
+        }
+        default: {
+            emitter = new BufferAsyncEmitter<T>(t, RxRingBuffer.SIZE);
+            break;
+        }
+        }
+
+        t.add(emitter);
+        t.setProducer(emitter);
+        asyncEmitter.call(emitter);
+        
+    }
+    
+    static final class CancellableSubscription 
+    extends AtomicReference<AsyncEmitter.Cancellable>
+    implements Subscription {
+        
+        /** */
+        private static final long serialVersionUID = 5718521705281392066L;
+
+        public CancellableSubscription(AsyncEmitter.Cancellable cancellable) {
+            super(cancellable);
+        }
+        
+        @Override
+        public boolean isUnsubscribed() {
+            return get() == null;
+        }
+        
+        @Override
+        public void unsubscribe() {
+            if (get() != null) {
+                AsyncEmitter.Cancellable c = getAndSet(null);
+                if (c != null) {
+                    try {
+                        c.cancel();
+                    } catch (Exception ex) {
+                        Exceptions.throwIfFatal(ex);
+                        RxJavaHooks.onError(ex);
+                    }
+                }
+            }
+        }
+    }
+    
+    static abstract class BaseAsyncEmitter<T> implements AsyncEmitter<T>, Producer, Subscription {
+        final Subscriber<? super T> actual;
+        
+        final AtomicLong requested;
+        
+        final SerialSubscription serial;
```

Comment:
Looks like you could extend from one of these two to continue saving a few bytes!